nano/node/fair_queue.hpp

```diff
@@ -0,0 +1,289 @@
+#pragma once
+
+#include <nano/lib/utility.hpp>
+#include <nano/node/transport/channel.hpp>
+
+#include <algorithm>
+#include <chrono>
+#include <deque>
+#include <functional>
+#include <memory>
+#include <numeric>
+#include <tuple>
+#include <utility>
+
+namespace nano
+{
+template <typename Request, typename Source>
+class fair_queue final
+{
+public:
+	/**
+	 * Holds user supplied source type(s) and an optional channel. This is used to uniquely identify and categorize the source of a request.
+	 */
+	struct origin
+	{
+		Source source;
+		std::shared_ptr<nano::transport::channel> channel;
+
+		origin (Source source, std::shared_ptr<nano::transport::channel> channel = nullptr) :
+			source{ source },
+			channel{ std::move (channel) }
+		{
+		}
+
+		bool alive () const
+		{
+			if (channel)
+			{
+				return channel->alive ();
+			}
+			// Some sources (eg. local RPC) don't have an associated channel, never remove their queue
+			return true;
+		}
+
+		auto operator<=> (origin const &) const = default;
+	};
+
+private:
+	struct entry
+	{
+		using queue_t = std::deque<Request>;
+		queue_t requests;
+
+		size_t priority;
```

Comment:
Priority seems like a misleading term because blocks with highest priority do not necessarily go first.
Throughput factor or queue width or something indicating heavier flow might be a better name.
I do not feel very strongly about this, feel free to ignore if you disagree.

Reply:
I'm happy to change it to a better name, though for aesthetic reasons it must be a single word.

Reply:
`allowance` or `quota`?